Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,220 @@
+every partial function `g`.
+-/
+theorem partrec_iff_partrec_in_everything (f : ℕ →. ℕ) : Nat.Partrec f ↔ ∀ g, f ≤ᵀ g :=
+  ⟨(·.turingReducible), (· _ |>.partrec_of_zero)⟩
+
+theorem TuringReducible.refl (f : ℕ →. ℕ) : f ≤ᵀ f := by
+  apply RecursiveIn.oracle; simp
+
+instance : IsRefl (ℕ →. ℕ) TuringReducible :=
+  ⟨fun f => TuringReducible.refl f⟩
+
+theorem TuringReducible.trans {f g h : ℕ →. ℕ} (hg : f ≤ᵀ g) (hh : g ≤ᵀ h) :
+    f ≤ᵀ h := by
+  induction' hg with g' hg g' h' _ _ ih₁ ih₂ g' h' _ _ ih₁ ih₂ g' h' _ _ ih₁ ih₂ g' _ ih
+  repeat {constructor}
+  · rw [Set.mem_singleton_iff] at hg; rw [hg]; exact hh
+  · case pair =>
+    apply RecursiveIn.pair ih₁ ih₂
+  · case comp =>
+    apply RecursiveIn.comp ih₁ ih₂
+  · case prec =>
+    apply RecursiveIn.prec ih₁ ih₂
+  · case rfind =>
+    apply RecursiveIn.rfind ih
+
+instance : IsTrans (ℕ →. ℕ) TuringReducible :=
+  ⟨@TuringReducible.trans⟩
+
+instance : IsPreorder (ℕ →. ℕ) TuringReducible where
+  refl := TuringReducible.refl
+
+theorem TuringEquivalent.equivalence : Equivalence TuringEquivalent :=
+  (AntisymmRel.setoid _ _).iseqv
+
+@[refl]
+theorem TuringEquivalent.refl (f : ℕ →. ℕ) : f ≡ᵀ f :=
+  Equivalence.refl equivalence f
+
+@[symm]
+theorem TuringEquivalent.symm {f g : ℕ →. ℕ} (h : f ≡ᵀ g) : g ≡ᵀ f :=
+  Equivalence.symm equivalence h
+
+@[trans]
+theorem TuringEquivalent.trans (f g h : ℕ →. ℕ) (h₁ : f ≡ᵀ g) (h₂ : g ≡ᵀ h) : f ≡ᵀ h :=
+  Equivalence.trans equivalence h₁ h₂
+
+/--
+Instance declaring that `RecursiveIn` is a preorder.
+-/
+instance : IsPreorder (ℕ →. ℕ) TuringReducible where
+  refl := TuringReducible.refl
+  trans := @TuringReducible.trans
+
+/--
+Turing degrees are the equivalence classes of partial functions under Turing equivalence.
+-/
+abbrev TuringDegree :=
+  Antisymmetrization _ TuringReducible
+
+instance TuringDegree.isPartialOrder : PartialOrder TuringDegree :=
+  @instPartialOrderAntisymmetrization (ℕ →. ℕ)
+    {le := TuringReducible, le_refl := TuringReducible.refl, le_trans := @TuringReducible.trans}
+
+lemma partrec_iff_partrec_in_empty (f : ℕ →. ℕ) : Nat.Partrec f ↔ RecursiveIn {} f := by
+  constructor
+  intros pF
+  · induction' pF with f' g' _ _ ih₁ ih₂ f' g' _ _ ih₁ ih₂ f' g' _ _ ih₁ ih₂ f' _ ih
+    repeat {constructor}
+    · case pair =>
+      apply RecursiveIn.pair ih₁ ih₂
+    · case comp =>
+      apply RecursiveIn.comp ih₁ ih₂
+    · case prec =>
+      apply RecursiveIn.prec ih₁ ih₂
+    · case rfind =>
+      apply RecursiveIn.rfind ih
+  · intro fRecInNone
+    induction' fRecInNone with g hg g h _ _ ih₁ ih₂ g h _ _ ih₁ ih₂ g h _ _ ih₁ ih₂ g _ ih
+    repeat {constructor}
+    · simp at hg
+    repeat {constructor; assumption; try assumption}
+
+/--
+An alternative definition of partial recursive in terms of oracle computability:
+A partial recursive function is a function which is recursive in the empty set
+-/
+def Partrec₃ (f : ℕ →. ℕ) : Prop :=
+  RecursiveIn {} f
```

Comment:
A definition without API is no good. Do you really think we should have this? Maybe you could make it an `example` instead:
```suggestion
/--
An alternative definition of partial recursive in terms of oracle computability:
A partial recursive function is a function which is recursive in the empty set
-/
example (f : ℕ →. ℕ) : Prop :=
  RecursiveIn {} f
```